crates/matrix-sdk-ui/src/timeline/tests/virt.rs

```diff
@@ -83,68 +84,107 @@ async fn test_day_divider() {
 
     // Nothing should happen if the fully read event isn't found.
     timeline.controller.handle_fully_read_marker(event_id!("$fake_event_id").to_owned()).await;
+    assert!(stream.next().now_or_never().is_none());
 
     // Nothing should happen if the fully read event is referring to an event
     // that has already been marked as fully read.
-    timeline.controller.handle_fully_read_marker(second_event_id).await;
+    timeline.controller.handle_fully_read_marker(event_id2).await;
+    assert!(stream.next().now_or_never().is_none());
 
-    timeline.handle_live_event(f.text_msg("D").sender(&ALICE)).await;
+    timeline.handle_live_event(f.text_msg("D").sender(&BOB)).await;
     let item = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
-    item.as_event().unwrap();
+    let event_id4 = item.as_event().unwrap().event_id().unwrap().to_owned();
 
-    timeline.controller.handle_fully_read_marker(third_event_id).await;
+    timeline.controller.handle_fully_read_marker(event_id3).await;
 
-    // The read marker is moved after the third event.
+    // The read marker is moved after the third event (sent by another user).
     assert_next_matches!(stream, VectorDiff::Remove { index: 3 });
     let marker = assert_next_matches!(stream, VectorDiff::Insert { index: 4, value } => value);
-    assert_matches!(marker.kind, TimelineItemKind::Virtual(VirtualTimelineItem::ReadMarker));
+    assert!(marker.is_read_marker());
+
+    // If the current user sends an event afterwards, the read marker doesn't move.
+    timeline.handle_live_event(f.text_msg("E").sender(&own_user)).await;
+    let item = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
+    item.as_event().unwrap();
+
+    assert!(stream.next().now_or_never().is_none());
+
+    // If the marker moved forward to another user's event, and there's no other
+    // event sent from another user, then it will be removed.
+    timeline.controller.handle_fully_read_marker(event_id4).await;
+    assert_next_matches!(stream, VectorDiff::Remove { index: 4 });
+
+    assert!(stream.next().now_or_never().is_none());
+
+    // When a last event is inserted by ourselves, still no read marker.
+    timeline.handle_live_event(f.text_msg("F").sender(&own_user)).await;
+    let item = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
+    item.as_event().unwrap();
+
+    timeline.handle_live_event(f.text_msg("G").sender(&own_user)).await;
+    let item = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
+    item.as_event().unwrap();
+
+    assert!(stream.next().now_or_never().is_none());
+
+    // But when it's another user who sent the event, then we get a read marker for
+    // their message.
+    timeline.handle_live_event(f.text_msg("H").sender(&BOB)).await;
+    let item = assert_next_matches!(stream, VectorDiff::PushBack { value } => value);
+    item.as_event().unwrap();
+
```

Comment:
Bear with me, just trying to understand the logic here:
* we left the read marker at `event_id4`
* we send 2 messages of our own and 1 from bob
* we get a read marker for `event_id8`?

Shouldn't the market have stayed at `event_id4` seeing how we never `handle_fully_read_marker`-ed it again?

Reply:
If the read marker was at `event_id4`, then it would look like the two events we sent ourselves are *before* the read marker, and this PR makes it so that events we sent ourselves are considered read, so should be *after* the read marker. I can beef up the code with comments showing what's displayed on the timeline at any interesting statement, if you think that can help with understanding the test code?

Reply:
Expanded the comments to show this in an hopefully clearer way.

Reply:
> I can beef up the code with comments

Thank you, it's super super clear now! 🙌